### PR TITLE
feat(bulk-import): use react query to fetch repositories

### PR DESCRIPTION
**Pull request by debsmita1**
**Resolves:**
https://issues.redhat.com/browse/RHIDP-4035

### CODE REVIEW

**Comment by its-mitesh-kumar** [comment]
@debsmita1 Could you please resolve quality gate failing and merge the latest change from main .

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.test.tsx [hunk 1]
```diff
@@ -97,7 +109,9 @@ describe('AddRepsositoriesForm', () => {
             ],
           ]}
         >
-          <AddRepositoriesForm error={null} />
+          <QueryClientProvider client={queryClient}>
+            <AddRepositories error={null} />
```
Comment (christoph-jerolimov) [inline plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.test.tsx:113]: I see that it's already before like this, but can you make the `error` prop optional and remove the prop here?

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx [hunk 2]
```diff
@@ -1,133 +1,112 @@
...
-    paddingLeft: '5px',
-  },
-}));
+export const AddRepositoriesForm = () => {
+  const bulkImportApi = useApi(bulkImportApiRef);
+  const navigate = useNavigate();
+  const [generalSubmitError, setGeneralSubmitError] = React.useState<{
+    message: string;
+    title: string;
+  } | null>(null);
```
Comment (christoph-jerolimov) [inline plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx:32]: You can use the error from the useMutation instead.

```suggestion
```

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx [hunk 3]
```diff
@@ -1,133 +1,112 @@
...
+      setGeneralSubmitError({
+        message:
+          error?.message ||
+          `${get(error, 'error.message')}\n${get(error, 'error.response.url')}` ||
+          'Failed to create pull request',
+        title: error?.name || get(error, 'error.name') || 'Error occured',
+      });
+      return error;
+    },
+  });
```
Comment (christoph-jerolimov) [inline plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx:63]: I think it's easier if you handle this later by using the mutationCreate.error

```suggestion
  const mutationCreate = useMutation(createImportJobs);
```

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx [hunk 4]
```diff
@@ -1,133 +1,112 @@
...
-        />
-      )}
-    </>
+    <DrawerContextProvider>
+      <Formik
+        initialValues={initialValues}
+        enableReinitialize
+        onSubmit={handleSubmit}
+      >
+        <AddRepositories error={generalSubmitError} />
```
Comment (christoph-jerolimov) [inline plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx:108]: Maybe you can pass just `mutationCreate.error` here. Or you can convert after `const mutationCreate` with

```tsx
  const generalSubmitError = mutationCreaten.errro ? ({
    message:
      error?.message ||
        `${get(error, 'error.message')}\n${get(error, 'error.response.url')}` ||
        'Failed to create pull request',
    title: error?.name || get(error, 'error.name') || 'Error occured',
  }) : null;
```

```suggestion
        <AddRepositories error={generalSubmitError} />
```

Wdyt?

File: plugins/bulk-import/src/hooks/useRepositories.test.ts [hunk 5]
```diff
@@ -10,12 +9,18 @@ jest.mock('@backstage/core-plugin-api', () => ({
...
+}));
 
 describe('useRepositories', () => {
   it('should return repositories', async () => {
-    mockUseApi.mockReturnValue({
-      dataFetcher: jest.fn().mockReturnValue(mockGetRepositories),
+    (useQuery as jest.Mock).mockReturnValue({
+      data: mockGetRepositories,
+      isLoading: false,
+      error: '',
```
Comment (christoph-jerolimov) [inline plugins/bulk-import/src/hooks/useRepositories.test.ts:22]: I would expect error is null when there is no error.

```suggestion
      error: null,
```

File: plugins/bulk-import/src/hooks/useRepositories.test.ts [hunk 6]
```diff
@@ -32,8 +37,11 @@ describe('useRepositories', () => {
   });
 
   it('should return organizations', async () => {
-    mockUseApi.mockReturnValue({
-      dataFetcher: jest.fn().mockReturnValue(mockGetOrganizations),
+    (useQuery as jest.Mock).mockReturnValue({
+      data: mockGetOrganizations,
+      isLoading: false,
+      error: '',
```
Comment (christoph-jerolimov) [inline plugins/bulk-import/src/hooks/useRepositories.test.ts:43]:
```suggestion
      error: null,
```

File: plugins/bulk-import/src/hooks/useRepositories.test.ts [hunk 7]
```diff
@@ -47,13 +55,16 @@ describe('useRepositories', () => {
...
+    (useQuery as jest.Mock).mockReturnValue({
+      data: {
         ...mockGetRepositories,
         repositories: mockGetRepositories.repositories?.filter(
           r => r.organization === 'org/dessert',
         ),
-      }),
+      },
+      isLoading: false,
+      error: '',
```
Comment (christoph-jerolimov) [inline plugins/bulk-import/src/hooks/useRepositories.test.ts:66]:
```suggestion
      error: null,
```

File: plugins/bulk-import/src/hooks/useRepositories.ts [hunk 8]
```diff
@@ -182,8 +191,8 @@ export const useRepositories = (options: {
   const debouncedUpdateResources = useDebounceCallback(prepareData, 200);
 
   React.useEffect(() => {
-    debouncedUpdateResources?.(value, loading);
-  }, [debouncedUpdateResources, value, loading]);
+    debouncedUpdateResources?.(value, isQueryLoading);
+  }, [debouncedUpdateResources, value, isQueryLoading]);
 
   return useDeepCompareMemoize({
```
Comment (christoph-jerolimov) [inline plugins/bulk-import/src/hooks/useRepositories.ts:197]: I see an issue that the pagination forward and backend shows always a loading indicator.

I guess that the `useRepository` doesn't use the caching benefits of react-query yet/correctly.

My recommendation is that you remove almost all useStates from this hook.

If you need calculate results, based on the useQuery hook, you can use `useMemo` instead.

The code should also get rid off the effect to call debouncedUpdateResources and the deepCompareMemoize function.

The order should be:

1. useQuery to fetch data
2. useMemo if you must transform "complex" data with a dependency just on the data.

**Review by christoph-jerolimov — CHANGES_REQUESTED** [review]
Hi, from the user perspective I see two issues:

1. When I paginate between the pages, I always see a loading indicator.
2. The order by feature works just on the current page. I expect that this issue exists before, but server-side pagination works only when the server also supports server-side ordering.

**Comment by debsmita1** [comment]
Created PR https://github.com/redhat-developer/rhdh-plugins/pull/23 , so closing this one

**Pull request closed without merging**